src/snowflake/connector/auth/by_plugin.py

```diff
@@ -61,32 +75,113 @@ def reset(self):
         self._current_sleep_time = 1
 
 
-class AuthByPlugin:
+@unique
+class AuthType(Enum):
+    DEFAULT = "SNOWFLAKE"  # default authenticator name
+    EXTERNAL_BROWSER = "EXTERNALBROWSER"
+    KEY_PAIR = "SNOWFLAKE_JWT"
+    OAUTH = "OAUTH"
+    ID_TOKEN = "ID_TOKEN"
+    USR_PWD_MFA = "USERNAME_PASSWORD_MFA"
+    OKTA = "OKTA"
+
+
+class AuthByPlugin(ABC):
     """External Authenticator interface."""
 
     def __init__(self) -> None:
         self._retry_ctx = AuthRetryCtx()
+        self.consent_cache_id_token = False
+        self._timeout = 120
+
+    @property
+    def timeout(self) -> int:
+        return self._timeout
+
+    @timeout.setter
+    def timeout(self, value: Any) -> None:
+        self._timeout = int(value)
 
     @property
-    def assertion_content(self):
+    @abstractmethod
+    def type_(self) -> AuthType:
+        """Return the Snowflake friendly name of auth class."""
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def assertion_content(self) -> str:
+        """Return a safe version of the information used to authenticate with Snowflake.
+
+        This is used for logging, useful for printing temporary tokens, but make sure to
+        mask secrets.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def prepare(
+        self,
+        *,
+        conn: SnowflakeConnection,
```

Comment:
The `AuthByKeyPair. prepare` does not take conn, should we set conn to have a default value `None`?
```suggestion
        conn: SnowflakeConnection | None = None,
```

Reply:
While some overwriting functions don't expect `conn` `**kwargs` was added to all of those `prepare` functions. And `SnowflakeConnection` will always pass itself in to `prepare`